### ISSUE TITLE
perfomance issue

### ISSUE THREAD

**Issue by fabriziomoscon** [tb0hdan/CSVDemo#6]
Do you see any performance issue with this line? https://github.com/tb0hdan/CSVDemo/blob/master/src/csvdemo/parse.go#L30
How would you refactor your code to be more performant?

**Comment by tb0hdan** [comment]
sendGRPCRequest is a blocking call that is executed sequentially for each CSV line. Even if DB latency is low, it can take forever to process. That's why simple worker pool (without timeouts or result checking) was added. It uses per CPU-core based approach (which can be adjusted according to requirements, e.g. fixed amount of connections).

**Comment by fabriziomoscon** [comment]
@tb0hdan I see your change, but I wanted you also to focus on the fact that you initialise a gRPC connection per each record. Is there a way to re-use a single gRPC connection? Or is there any particular way to do in the way your test is doing it?

**Comment by tb0hdan** [comment]
Indeed. Moving out Dial call ( https://github.com/tb0hdan/CSVDemo/blob/master/src/csvdemo/main.go#L80 ) from the function and reorganizing it as a structure + bound method will allow connection initialization on a per-worker basis. Then, if worker error occurs (i.e. connection is closed), it will be restarted and connection re-established.  There are some other techniques like connection pooling (through external library) and load balancing, though I haven't tried them (yet).

**Comment by fabriziomoscon** [comment]
thanks for the explanation
